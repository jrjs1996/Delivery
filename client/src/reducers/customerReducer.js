import {
  ADD_CUSTOMER,
  DELETE_CUSTOMER,
  FETCH_CUSTOMERS,
  GET_CUSTOMER,
  UPDATE_CUSTOMER,
} from '../actions/types';
import { updateItem, deleteItem, insertItem, createItems } from './utils';

const initialState = {
  items: {},
  item: {},
};

export default function (state = initialState, action) {
  switch (action.type) {
    case FETCH_CUSTOMERS:
      return {
        ...state,
        items: createItems(action.payload, '_id'),
      };
    case ADD_CUSTOMER:
      return {
        ...state,
        items: insertItem(state.items, action.payload, '_id'),
      };
    case GET_CUSTOMER:
      return {
        ...state,
        item: action.payload,
      };
    case UPDATE_CUSTOMER:
      return {
        ...state,
        items: updateItem(state.items, action.payload, '_id'),
      };
    case DELETE_CUSTOMER:
      return {
        ...state,
        items: deleteItem(state.items, action.payload, '_id'),
      };
    default:
      return state;
  }
}
