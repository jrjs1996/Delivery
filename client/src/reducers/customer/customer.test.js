import customerReducer from './customer';
import {
  ADD_CUSTOMER,
  DELETE_CUSTOMER,
  FETCH_CUSTOMERS,
  UPDATE_CURRENT_CUSTOMER,
  UPDATE_CUSTOMER,
} from '../../actions/types';
import { customersMock } from '../../tests/mocks';

const createState = (items, currentCustomer) => ({
  items,
  currentCustomer,
});

const createAction = (type, payload) => ({
  payload,
  type,
});

describe('Customer reducer', () => {
  it('Initial state is empty', () => {
    const initialState = customerReducer(undefined, 'FAKE_ACTION');
    expect(initialState).toMatchSnapshot();
  });

  it("Returns given state when an action doesn't match", () => {
    const state = customerReducer({ mockState: 'MockState' }, 'FAKE_ACTION');
    expect(state).toMatchSnapshot();
  });

  it('Adds customer when calling ADD_CUSTOMER', () => {
    const initialItems = { [customersMock[0]._id]: customersMock[0] };
    const state = createState(initialItems, undefined);
    const action = createAction(ADD_CUSTOMER, customersMock[1]);
    expect(customerReducer(state, action)).toMatchSnapshot();
  });

  it('Removes customer when calling DELETE_CUSTOMER', () => {
    const initialItems = {
      [customersMock[0]._id]: customersMock[0],
      [customersMock[1]._id]: customersMock[1],
    };
    const state = createState(initialItems, undefined);
    const action = createAction(DELETE_CUSTOMER, customersMock[0]);
    expect(customerReducer(state, action)).toMatchSnapshot();
  });

  it('Populates items when calling FETCH_CUSTOMERS', () => {
    const action = createAction(FETCH_CUSTOMERS, customersMock);
    expect(customerReducer(createState(), action)).toMatchSnapshot();
  });

  it('Updates current customer to be payload when calling UPDATE__CURRENT_CUSTOMER', () => {
    const state = createState(undefined, { name: 'InitialCurrentCustomer' });
    const action = createAction(UPDATE_CURRENT_CUSTOMER, { name: 'NewCurrentCustomer' });
    expect(customerReducer(state, action)).toMatchSnapshot();
  });

  it('Updates given customer in items to be payload when calling UPDATE_CUSTOMER', () => {
    const initialItems = {
      [customersMock[0]._id]: customersMock[0],
      [customersMock[1]._id]: customersMock[1],
    };
    const state = createState(initialItems, undefined);
    const action = createAction(UPDATE_CUSTOMER, {
      _id: customersMock[0]._id,
      firstName: 'UPDATED INFO',
    });
    expect(customerReducer(state, action)).toMatchSnapshot();
  });
});
