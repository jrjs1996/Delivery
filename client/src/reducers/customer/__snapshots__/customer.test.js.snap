// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customer reducer Adds customer when calling ADD_CUSTOMER 1`] = `
Object {
  "currentCustomer": undefined,
  "items": Object {
    "0": Object {
      "__v": 0,
      "_id": "0",
      "addresses": Array [
        "123 green st",
        "232 8th St E",
      ],
      "email": "johnD@email.com",
      "firstName": "John",
      "lastName": "Davidson",
    },
    "1": Object {
      "__v": 0,
      "_id": "1",
      "addresses": Array [
        "488 5th Ave",
      ],
      "email": "Bobby@email.com",
      "firstName": "Bob",
      "lastName": "Harris",
    },
  },
}
`;

exports[`Customer reducer Initial state is empty 1`] = `
Object {
  "currentCustomer": Object {},
  "items": Object {},
}
`;

exports[`Customer reducer Populates items when calling FETCH_CUSTOMERS 1`] = `
Object {
  "currentCustomer": undefined,
  "items": Object {
    "0": Object {
      "__v": 0,
      "_id": "0",
      "addresses": Array [
        "123 green st",
        "232 8th St E",
      ],
      "email": "johnD@email.com",
      "firstName": "John",
      "lastName": "Davidson",
    },
    "1": Object {
      "__v": 0,
      "_id": "1",
      "addresses": Array [
        "488 5th Ave",
      ],
      "email": "Bobby@email.com",
      "firstName": "Bob",
      "lastName": "Harris",
    },
  },
}
`;

exports[`Customer reducer Removes customer when calling DELETE_CUSTOMER 1`] = `
Object {
  "currentCustomer": undefined,
  "items": Object {
    "1": Object {
      "__v": 0,
      "_id": "1",
      "addresses": Array [
        "488 5th Ave",
      ],
      "email": "Bobby@email.com",
      "firstName": "Bob",
      "lastName": "Harris",
    },
  },
}
`;

exports[`Customer reducer Returns given state when an action doesn't match 1`] = `
Object {
  "mockState": "MockState",
}
`;

exports[`Customer reducer Updates current customer to be payload when calling UPDATE__CURRENT_CUSTOMER 1`] = `
Object {
  "currentCustomer": Object {
    "name": "NewCurrentCustomer",
  },
  "items": undefined,
}
`;

exports[`Customer reducer Updates given customer in items to be payload when calling UPDATE_CUSTOMER 1`] = `
Object {
  "currentCustomer": undefined,
  "items": Object {
    "0": Object {
      "__v": 0,
      "_id": "0",
      "addresses": Array [
        "123 green st",
        "232 8th St E",
      ],
      "email": "johnD@email.com",
      "firstName": "UPDATED INFO",
      "lastName": "Davidson",
    },
    "1": Object {
      "__v": 0,
      "_id": "1",
      "addresses": Array [
        "488 5th Ave",
      ],
      "email": "Bobby@email.com",
      "firstName": "Bob",
      "lastName": "Harris",
    },
  },
}
`;
